src/seq.rs

```diff
@@ -110,116 +117,204 @@ pub fn sample_slice<R, T>(rng: &mut R, slice: &[T], amount: usize) -> Vec<T>
-/// have the indices themselves so this is provided as an alternative.
+/// The implementation used is not specified; we automatically select the
+/// fastest available implementation for the `length` and `amount` parameters
+/// (based on detailed profiling on an Intel Haswell CPU). Roughly speaking,
+/// complexity is `O(amount)`, except that when `amount` is small, performance
+/// is closer to `O(amount^2)`, and when `length` is close to `amount` then
+/// `O(length)`.
 ///
-/// Panics if `amount > length`
-pub fn sample_indices<R>(rng: &mut R, length: usize, amount: usize) -> Vec<usize>
+/// Note that we only support `u32` indices since this covers the vast majority
+/// of uses, and performance is significantly better than with `u64`.
+/// 
+/// If an allocation-free `no_std` function is required, it is suggested
+/// to adapt the internal `sample_indices_floyd` implementation.
+///
+/// Panics if `amount > length` or if `length` is not reprentable as a `u32`.
+pub fn sample_indices<R>(rng: &mut R, length: usize, amount: usize,
+    shuffled: bool) -> Vec<u32>
     where R: Rng + ?Sized,
 {
     if amount > length {
-        panic!("`amount` must be less than or equal to `slice.len()`");
+        panic!("`amount` of samples must be less than or equal to `length`");
     }
-
-    // We are going to have to allocate at least `amount` for the output no matter what. However,
-    // if we use the `cached` version we will have to allocate `amount` as a HashMap as well since
-    // it inserts an element for every loop.
-    //
-    // Therefore, if `amount >= length / 2` then inplace will be both faster and use less memory.
-    // In fact, benchmarks show the inplace version is faster for length up to about 20 times
-    // faster than amount.
-    //
-    // TODO: there is probably even more fine-tuning that can be done here since
-    // `HashMap::with_capacity(amount)` probably allocates more than `amount` in practice,
-    // and a trade off could probably be made between memory/cpu, since hashmap operations
-    // are slower than array index swapping.
-    if amount >= length / 20 {
-        sample_indices_inplace(rng, length, amount)
+    if length > (::core::u32::MAX as usize) {
+        panic!("`length` is not representable as `u32`");
+    }
+    let amount = amount as u32;
+    let length = length as u32;
+    
+    // Choice of algorithm here depends on both length and amount. See:
+    // https://github.com/rust-lang-nursery/rand/pull/479
+    // We do some calculations with u64 to avoid overflow.
+
+    if amount < 442 {
+        const C: [[u64; 2]; 2] = [[5, 45], [50, 350]];
+        let j = if length < 500_000 { 0 } else { 1 };
+        let m4 = 6 * amount as u64;
+        if C[0][j] * (length as u64) < (C[1][j] + m4) * amount as u64 {
+            sample_indices_inplace(rng, length, amount)
+        } else if shuffled {
+            sample_indices_floyd_shuffled(rng, length, amount)
```

Comment:
I like the permutation and shuffling algorithm. It really uses the minimum number of values from the RNG possible. But in this case using `shuffle` is just faster:
```rust
let mut indices = sample_indices_floyd(rng, length, amount);
rng.shuffle(&mut indices);
indices
```

Reply:
Hmm, I guess it could be. I should have benchmarked that!